### PR TITLE
EDUCATOR-854: Update API client to support pagination, filtering, and sorting of course_summaries

### PR DESCRIPTION
**Pull request by kdmccormick**
[This](https://openedx.atlassian.net/wiki/spaces/AC/pages/161994283/Analytics+Data+API+-+Paginated+Course+Summaries) document (note: that document is not currently publicly accessible; we need to change that) describes a set of changes to the Analytics Data API. In short, the `course_summaries/` endpoint is gaining pagination, filtering, and sorting, and the `course_aggregate_data/` endpoint has been added.

This PR introduces support for those changes into the API Client. Additionally, I have refactored some of the code and tests to best accommodate the new functionality.

### Note:
* This should be merged AFTER the Analytics API PR.
* Two of the tests relating to [special characters in parameters](https://github.com/edx/edx-analytics-data-api-client/pull/36/files#diff-72c934fcb031db093dce56482ec30303R36) were failing. For now, I have changed them to not to use the `'+'` character. This seems to be because of a problem with `httpretty`. Because we know that the `'+'` character doesn't cause a problem in practice, I'm going to leave the test as is.

### Links:
* JIRA
  * [Ticket](https://openedx.atlassian.net/browse/EDUCATOR-854)
  * [Epic](https://openedx.atlassian.net/browse/EDUCATOR-847)
* Analytics Data API
  * [Prototype PR](https://github.com/edx/edx-analytics-data-api/pull/177)
  * [Final PR](https://github.com/edx/edx-analytics-data-api/pull/178) (WIP)
* Analytics Data API Client
  * [Prototype PR](https://github.com/edx/edx-analytics-data-api-client/pull/35)
  * __Final PR__ (this)
* Insights
  * [Prototype/Final PR](https://github.com/edx/edx-analytics-dashboard/pull/710)

@edx/educator-dahlia

### CODE REVIEW

File: analyticsclient/_base.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import analyticsclient.constants.data_format as DF
```
Comment (efischer19) [inline analyticsclient/_base.py]: Nit: `DF` is not terribly descriptive, can you name this `DATA_FORMAT` or something similar?

Question: why does this file contain a leading `_`? That feels unnecessary.

File: analyticsclient/client.py [hunk 2]
```diff
@@ -46,41 +45,32 @@ def __init__(self, base_url, auth_token=None, timeout=0.25):
...
         """
         Retrieve the data for a resource.
 
-        Arguments:
-
-            resource (str): Path in the form of slash separated strings.
-            timeout (float): Continue to attempt to retrieve a resource for this many seconds before giving up and
-                raising an error.
-            data_format (str): Format in which data should be returned
+        Equivalent to `request(http_methods.GET, ...)`
```
Comment (efischer19) [inline analyticsclient/client.py]: I really like this change, good refactor

File: analyticsclient/status.py [hunk 3]
```diff
@@ -1,20 +1,10 @@
+from analyticsclient._base import BaseEndpoint
 from analyticsclient.exceptions import ClientError
 
 
-class Status(object):
+class Status(BaseEndpoint):
```
Comment (efischer19) [inline analyticsclient/status.py]: Really digging the base class refactor a lot after seeing all these repeated `__init__` methods vanish

**Comment by kdmccormick** [comment]
Thanks for the review @efischer19 ! Will make your suggested changes.

File: analyticsclient/course_summaries.py [hunk 4]
```diff
@@ -1,36 +1,63 @@
...
+
         Arguments:
-            course_ids: Array of course IDs as strings to return.  Default is to return all.
-            fields: Array of fields to return.  Default is to return all.
-            exclude: Array of fields to exclude from response. Default is to not exclude any fields.
-            programs: If included in the query parameters, will include the programs array in the response.
+            order_by (str): Summary field to sort by.
+            sort_order (str): Order of the sort.
+            course_ids (list[str]): Course IDs to filter by.
+            availability (list[str]) Availabilities to filter by.
```
Comment (thallada) [inline analyticsclient/course_summaries.py]: I think pacing_type is missing from this doc string

File: analyticsclient/tests/__init__.py [hunk 5]
```diff
@@ -49,67 +49,102 @@ def verify_last_querystring_equal(self, expected_query):
...
+        for key, val in expected_body.iteritems():
+            if not isinstance(val, list):
+                expected_body[key] = [val]
+        actual_body = httpretty.last_request().parsed_body
+        self.assertDictEqual(actual_body or {}, expected_body)
+
+    def test_request_with_many_ids(self):
+        """Endpoint can be called with a large number of ID parameters."""
+        params = self.fill_in_empty_params_with_dummies(**{self.id_field: ['id'] * 10000})
+        self.verify_post_data(**params)
```
Comment (thallada) [inline analyticsclient/tests/__init__.py]: I might just be missing the test somewhere else, but shouldn't this test that with a low number of ids (default < 10), that the endpoint does a GET instead?

File: analyticsclient/tests/test_course_summaries.py [hunk 6]
```diff
@@ -1,30 +1,53 @@
...
+        'sort_order',
+    ])
+    _INT_PARAMS = set([
+        'page',
+        'page_size',
+    ])
+    _ALL_PARAMS = _LIST_PARAMS | _STRING_PARAMS | _INT_PARAMS
+    other_params = _ALL_PARAMS
+
+    _TEST_STRING = 'a'  # For now, disable testing of URL encoding 'Aa1_-:/+* '  # Test URL encoding
```
Comment (thallada) [inline analyticsclient/tests/test_course_summaries.py]: Oh, does the url encoding not work presently?

File: analyticsclient/course_summaries.py [hunk 7]
```diff
@@ -1,36 +1,68 @@
...
+            program_ids (list[str]): Course IDs of programs to filter by.
+            text_search (str): Sub-string to search for in course titles and IDs.
+            page (int): Page number.
+            page_size (int): Size of page.
+            fields (list[str]) Fields of course summaries to return in response.
+            exclude (list[str]) Fields of course summaries to NOT return in response.
+            data_format (str): Data format for response. Must be data_format.JSON or
+                               data_format.CSV.
+            request_all (bool): Whether all summaries should be returned, or just a
+                                single page. Overrides `page` and `page_size`.
```
Comment (thallada) [inline analyticsclient/course_summaries.py]: I think `pacing_type` is missing from this list

Comment (kdmccormick) [inline analyticsclient/course_summaries.py]: Good catch!

File: analyticsclient/tests/__init__.py (hunk 5 above)
Comment (kdmccormick) [inline analyticsclient/tests/__init__.py]: [This](https://github.com/edx/edx-analytics-data-api-client/pull/36/files/2ffb6e64e8b7b18f748e0ddbf4625f03bda5ef21#diff-d8b9f032ba99ee92e0627861280e7fadR85) test takes care of that.

**Pull request merged**